Calendar: Recurrence: Rewrite logic to separate instances, exceptions and exclusions

benbucksch opened the pull request.

NeilRashbrook commented on app/logic/Calendar/Event.ts:
```diff
@@ -481,18 +536,17 @@ export class Event extends Observable {
     assert(this.calendar, "To delete an event, it needs to be in a calendar first");
     assert(this.calendar.storage, "To delete an event, the calendar needs to be saved first");
+    // this.recurrenceRule = null; is overkill for removing exceptions; the database already cascades the delete
+    this.calendar.events.remove(this.exceptions);
```
```suggestion
    this.calendar.events.removeAll(this.exceptions);
```

NeilRashbrook commented:
FYI when I delete an instance from the sidebar, the `remove` notifies an `AdditionCollectionDedupObserver` (whose `_mapFunc` is `mapToEvents`) which notifies another `AdditionCollectionDedupObserver` (whose `_mapFunc` is `cal => cal.eventsWithRecurrences`) but that one doesn't have anything observing it.

If I remove the `.sortBy(ev => ev.startTime)` then the delete happens instantly. It looks as if the observers are getting garbage collected due to `svelte-collections` version 1.1.23 switching to `WeakSetIterable` for its _observers. (Some observers are set as properties on their collection which stops them from getting garbage collected.)

benbucksch commented:
> It looks as if the observers are getting garbage collected due to svelte-collections version 1.1.23 switching to WeakSetIterable for its _observers.

Thank you! I would have hunted that bug for quite a while.
Indeed, given that most observers are only set as observers to the collection, it makes no sense to make them weak refs. That was a bad change. I've reverted that, released it as svelte-collections 1.2.11, and added it to the branch.

benbucksch commented on app/logic/Calendar/SQL/SQLEvent.ts:
```diff
@@ -156,15 +155,15 @@ export class SQLEvent extends Event {
+        parentEvent.fillRecurrences();
       }
     }
     if (row.recurrenceRule) {
```
I'm now calling generate() after both lines. That should fix it, right?

NeilRashbrook commented on app/logic/Calendar/Event.ts:
```diff
@@ -153,12 +153,12 @@ export class Event extends Observable {
-    this._recurrenceRule = rule;
-    this.recurrenceCase = RecurrenceCase.Master; // notifies
-    this.instances.clear();
+    this.generateRecurringInstances();
```
Oops, you're comparing the rule after you replace it...

benbucksch replied:
Fixed by Neil in #671
